src/main/c/NativeLongFunc.cpp

```diff
@@ -0,0 +1,137 @@
+// src/main/c/NativeFunc.cpp - native funcs for Scala NativeLongFunc
+#include "hail/NativePtr.h"
+#include <assert.h>
+#include <jni.h>
+
+namespace hail {
+
+namespace {
+
+using NativeFunc = NativeFuncObj<long>;
+
+NativeFunc* to_NativeFunc(JNIEnv* env, jobject thisJ) {
+  // Should be a dynamic_cast, but RTTI causes trouble
+  return reinterpret_cast<NativeFunc*>(get_from_NativePtr(env, thisJ));
```

Comment:
`NativeFuncObj<long>` is a subclass of `NativeObj`, so isn't static_cast what you want here?

Reply:
No, static_cast only works the other way round, 
 static_cast<BaseClass*>(expr of type SubClass*) is ok, because we know at compile
time that the cast is valid.

To get from BaseClass* to SubClass*, we need either 

  dynamic_cast<SubClass*>, which will check run-time type info and give us a nullptr
  if the object isn't of the expected type.

or reinterpret_cast<SubClass*>, which means "compiler can't tell what this is, but I know for
  sure it's ok to treat it as a SubClass*, so just do it".

Reply:
According to https://en.cppreference.com/w/cpp/language/static_cast, referring to `static_cast < new_type > ( expression )`
> If new_type is a pointer or reference to some class D and the type of expression is a pointer or reference to its non-virtual base B, `static_cast` performs a *downcast*. This downcast is ill-formed if B is ambiguous, inaccessible, or virtual base (or a base of a virtual base) of D. Such `static_cast` makes no runtime checks to ensure that the object's runtime type is actually D, and may only be used safely if this precondition is guaranteed by other means, such as when implementing static polymorphism. Safe downcast may be done with dynamic_cast.

In particular, if the baseclass doesn't live at offset 0 inside the subclass, then `static_cast` will perform the correct downcast, adjusting the pointer to the start of the subclass. On the other hand,
> Unlike static_cast, but like const_cast, the reinterpret_cast expression does not compile to any CPU instructions. It is purely a compiler directive which instructs the compiler to treat the sequence of bits (object representation) of expression as if it had the type new_type.
so `reinterpret_cast` won't do the pointer adjustment and will give undefined behavior in the case where the baseclass is not at offset 0.

I'm guessing we'll never have a case where the baseclass is not at offset 0, but it still seems to me that `static_cast` is the best choice here (given that we're avoiding RTTI).

Reply:
You are totally right again.  static_cast can be used for the downcast, and is better than
reinterpret_cast.  I'll fix it.